main/src/ca/uwaterloo/flix/api/lsp/provider/HoverProvider.scala

```diff
@@ -28,56 +30,42 @@ import scala.annotation.tailrec
+      override def consumeDefSymUse(symUse: DefSymUse): Unit = push(symUse)
+      override def consumeSigSymUse(symUse: SigSymUse): Unit = push(symUse)
+      override def consumeOpSymUse(symUse: OpSymUse): Unit = push(symUse)
+      override def consumeDef(defn: Def): Unit = push(defn)
+      override def consumeSig(sig: Sig): Unit = push(sig)
+      override def consumeOp(op: Op): Unit = push(op)
+      override def consumeFormalParam(fparam: FormalParam): Unit = push(fparam)
     }
-  }
-
-  @tailrec
-  private def hoverEntity(entity: Entity, uri: String, pos: Position)(implicit Index: Index, root: Root, flix: Flix): JObject = entity match {
-
-    case Entity.Case(caze) => hoverType(caze.tpe, caze.sym.loc)
-
-    case Entity.StructField(field) => hoverType(field.tpe, field.sym.loc)
-
-    case Entity.DefUse(sym, loc, _) => hoverDef(sym, loc)
-
-    case Entity.SigUse(sym, loc, _) => hoverSig(sym, loc)
-
-    case Entity.VarUse(_, _, parent) => hoverEntity(parent, uri, pos)
-
-    case Entity.CaseUse(_, _, parent) => hoverEntity(parent, uri, pos)
-
-    case Entity.StructFieldUse(_, _, parent) => hoverEntity(parent, uri, pos)
-
-    case Entity.Exp(exp) => hoverTypeAndEff(exp.tpe, exp.eff, exp.loc)
-
-    case Entity.FormalParam(fparam) => hoverType(fparam.tpe, fparam.loc)
-
-    case Entity.Pattern(pat) => hoverType(pat.tpe, pat.loc)
-
-    case Entity.Pred(pred, tpe) => hoverType(tpe, pred.loc)
-
-    case Entity.LocalVar(sym, tpe) => hoverType(tpe, sym.loc)
 
-    case Entity.Type(t) => hoverKind(t)
+    Visitor.visitRoot(root, HoverConsumer, Visitor.InsideAcceptor(uri, pos))
 
-    case Entity.OpUse(sym, loc, _) => hoverOp(sym, loc)
+    stack match {
+      case Nil => mkNotFound(uri, pos)
+      case head :: _ => hoverAny(head, uri, pos)
+    }
+  }
 
-    case Entity.Trait(_) => mkNotFound(uri, pos)
-    case Entity.Def(_) => mkNotFound(uri, pos)
-    case Entity.Effect(_) => mkNotFound(uri, pos)
-    case Entity.Enum(_) => mkNotFound(uri, pos)
-    case Entity.Struct(_) => mkNotFound(uri, pos)
-    case Entity.TypeAlias(_) => mkNotFound(uri, pos)
-    case Entity.AssocType(_) => mkNotFound(uri, pos)
-    case Entity.Label(_) => mkNotFound(uri, pos)
-    case Entity.Op(_) => mkNotFound(uri, pos)
-    case Entity.Sig(_) => mkNotFound(uri, pos)
-    case Entity.TypeVar(_) => mkNotFound(uri, pos)
+  private def hoverAny(x: AnyRef, uri: String, pos: Position)(implicit root: Root, flix: Flix): JObject = x match {
+    case tpe: Type => hoverKind(tpe)
+    case exp: Expr => hoverTypeAndEff(exp.tpe, exp.eff, exp.loc)
+    case DefSymUse(sym, loc) => hoverDef(sym, loc)
+    case SigSymUse(sym, loc) => hoverSig(sym, loc)
+    case OpSymUse(symUse, loc) => hoverOp(symUse, loc)
+    case Def(sym, _, _, loc) => hoverDef(sym, loc)
+    case Sig(sym, _, _, loc) => hoverSig(sym, loc)
+    case Op(sym, _, loc) => hoverOp(sym, loc)
+    case FormalParam(_, _, tpe, _, loc) => hoverType(tpe, loc)
```

Comment:
Would this case not already be covered by the tpe: Type case? I thought that was part of the advantage of the visitor.

Reply:
Howering on a type and hovering on a formal param is not quite the same, though you might be right that there is little reason to support both. For some formal param `x: T`, hovering on `x` is hovering on the formal param, but not the type `T`. If we only support `tpe: Type`, then hovering on the `x` itself, would do nothing. So the FormalParam case does add *something*. Right now, hovering on `x: T` but *not* `T` would give the type of `x` (T itself), which is arguably redundant, as it's a repition of the information you're hovering on. Hovering on the `T` itself, gives the kind of `T`.

We could remove FormalParam, I think that might be reasonable to do, since it doesn't really add anything. I added it based on the current behaviour of the hover provider. But maybe I shouldn't base this new provider off of the behaviour of the old. Better to rethink it entirely, perhaps?